fedimint-client/src/oplog.rs

```diff
@@ -64,45 +67,107 @@ impl OperationLog {
         limit: usize,
         start_after: Option<ChronologicalOperationLogKey>,
     ) -> Vec<(ChronologicalOperationLogKey, OperationLogEntry)> {
+        const EPOCH_DURATION: Duration = Duration::from_secs(60 * 60 * 24 * 7);
+
         let mut dbtx = self.db.begin_transaction_nc().await;
-        let operations: Vec<ChronologicalOperationLogKey> = dbtx
-            .find_by_prefix_sorted_descending(&ChronologicalOperationLogKeyPrefix)
-            .await
-            .map(|(key, ())| key)
-            // FIXME: this is a schlemil-the-painter algorithm that will take longer the further
-            // back in history one goes. To avoid that I see two options:
-            //   1. Add a reference to the previous operation to each operation log entry,
-            //      essentially creating a linked list, which seem a little bit inelegant.
-            //   2. Add an option to prefix queries that allows to specify a start key
-            //
-            // The current implementation may also skip operations due to `SystemTime` not being
-            // guaranteed to be monotonous. The linked list approach would also fix that.
-            .skip_while(|key| {
-                let skip = if let Some(start_after) = start_after {
-                    key.creation_time >= start_after.creation_time
-                } else {
-                    false
-                };
-
-                std::future::ready(skip)
-            })
-            .take(limit)
-            .collect::<Vec<_>>()
-            .await;
+        let operation_log_keys = if let Some(start_after) = start_after {
+            let Some(last_key) = dbtx
```

Comment:
Would cache when `Some`. It can't ever change.

Reply:
Does it mean we just have to impl the same thing we implemented for event log (a single writer serializing the order)?

Reply:
> Would cache when Some. It can't ever change.

Is that really worth it? You aren't supposed to call this fn in a hot loop. Can add it in a separate commit and then we decide.

> Theoretically a new entry can get added with previous timestamp at any point, right?.

Sure, but that would also trigger some sort of event in the future I assume? We currently don't deal with the case "new operation was inserted, send diff/insert to frontend". Currently, you just have to fetch the log again.

In practice, DB transactions, that add operation log entries, aren't that long-running. So the time it takes a user to navigate from "send transaction" to "show transaction log" would typically be sufficient to commit the dbtx.

You might be thinking about it as an API to get updates, but for that it's entirely ill-suited because it goes backwards in time and not forwards. The event log is the right abstraction for that use case. `list_operations` is about historic data that gets lazy-loaded as you scroll through your transaction log.

Thinking about it, we might want to have a "forward" function too at some point, so a UI can drop more current data if a user scrolls too far back and re-fetch it when they scroll up again.

Reply:
All it means is that this method can't be used for reliably noticing new operations, as it can miss new entries unless queried from beginning, and probably we should explicitly state it in the documentation. But this is probably not an issue.

Reply:
> Is that really worth it?

:shrug: Low gain, low effort.